### ISSUE TITLE
linux4.19 ppc64-musl observations and related things (chatty)

### ISSUE THREAD

**Issue by pullmoll** [void-linux/void-packages#10326]
I'm probably as of now the only one seeing this problem, yet I wanted to ask if anyone has seen such errors before. Trying to boot a Void installation for `ppc64-musl` on a PowerMac G5 with a Void built kernel (4.19.x) fails with the following messages:
```
agpgart: disagrees about version info of symbol module_layout
```
The same error is printed for the modules `crc_itu_t`, `scsi_mod` and `usb_core`.
After a long time, and the *windpark* beginning to evoke a hurrican in the G5, the `initrd` presents a prompt and I'm stuck. I cannot debug any further because without USB I don't have a keyboard... duh!

I can, however, boot into Void using Debian Sid's `vmlinux-3.16.0-6-powerpc64` and `initrd-3.16.0-6.img` copied to Void's `/boot` and Sid's modules copied to `/usr/lib/modules/3.16.0-6-powerpc64`. In this case Debian is the chicken and Void is the egg it seems.

I also started to try to build Void's `linux3.16` package for `ppc64-musl` but the configuration and template is giving me headaches. I started from Debian Sid's `/boot/config-3.16.0-6-powerpc64` yet the template needs (a lot of) fixing also. Anyway, I suspect the same or similar problems would happen there as well because the root of the problem is elsewhere. And then I want to run an up to date kernel, not an LTS.

#### Other observations while trying to to run Void userland:

`dracut -f` segfaults when trying to call `dracut-install` with the parameter `-P *(kernel/fs/nfs|kernel/fs/nfsd|kernel/fs/lockd)*` which may be due to a bug in `musl-fts`... at least I think `dracut` uses the file system traversal functions for an expression like this.
I can work around it by e.g. running `filesystems="hfs ext4 msdos" dracut -f --kver 4.19.31_1 -f`. If `filesystems` is defined, the problematic call is not run by `/usr/bin/dracut`.

Yet another and much more annoying observation is that `libcrypto.so.44` fails hard. Programs using it get killed! No way to debug them with `gdb`, they just exit printing `Killed`. This happens for e.g. `ssh`, thus `rsync`, and of course every `https://...` connection of `curl` or `wget`, which currently makes testing *a bit cumbersome*. I need to copy e.g. current sources and the void-packages repo from my x86_64 to the G5 after I booted into Debian Sid and mounted the Void partitions... yay :)

Oh, BTW, trying to get `strace` built for `ppc64-musl` fails because `libunwind` would need a patchset for `ppc64-musl` similar to the one for `ppc-musl` which can be found [here](https://github.com/lede-project/source/commit/05dba65569756a2da28afb9fcfc4486105077cd0). It's all about `musl` not exposing the internal structure of the machine's context to the world, i.e. has an anonymous `ucontext_t`.

I then tried to build `libressl` on the G5 and it fails in 29 of the tests like `bntest` , `cipher_list`, `clienttest`, `configtest`, `dhtest`, `dsatest`, `ecdhtest`, ... All of them them just print `<PID> Killed` .
I also tried to build with `--disable-asm` to see if the problem is there, but the results are the same.

Of course I'm trying to build native, in Void userland, running a foreign kernel, but could that be a problem? So my question is does anyone have an idea where to look for the root of these problems?

A fun fact on the pro side: I can run Debian Sid binaries which are glibc like Debian's `mac-fdisk` after simply copying Debian's `libc-2.19.so` and `libc.so.6` symlink to Void's `/usr/lib`.

**Comment by newbluemoon** [comment]
Maybe having a look at how Adélie Linux builds its ppc64 kernel (and/or some other packages) can shed some light? That is if you haven’t done so already. ;) They’re at 4.14 though.

https://code.foxkit.us/adelie/packages/tree/master/system/easy-kernel

**Comment by pullmoll** [comment]
No, I did not yet look at Adélie Linux but will probably do that, too.

Currently I'm building a ppc64 glibc base system to verify that there really is a problem with the ABI elfv2 and powerpc64 big-endian. I don't see it yet. I could already build `base-chroot` and `base-system` without errors after creating a `cross-powerpc64-linux-gnu` package and adding the `common/cross-profiles/ppc64.sh` file.

But now it is time for some :beer: or :wine_glass:  to relax from that stuff.
Thank you for the hint!

**Comment by pullmoll** [comment]
New results FYI:
+ The problem with running current Void userland under a 3.16.x kernel was that this kernel lacks an implementation of `get_random()`. With the kernel returning `-1` for `get_random(buff, size)` and setting errno to `ENOSYS` many programs just fail hard.
+ I upgraded Debian to `unstable` (10.0), which itself uses `linux-4.19`, and it works just great with the ppc64 kernel and ppc userland. This includes a working `X` using the `nouveau` driver for video. I can also boot into a Void installation using this kernel and Debian's `initrd.img`.
+ I'm running Void installed from home grown packages for `ppc64` with `ABI=elfv2` and it works more or less. It can now run `sshd` and `ssh`, which makes updating a lot easier.
+ A main problem is that `glibc`'s `ldd` prints `not a dynamic executable` for any shared libraries and binaries, e.g. `ldd /usr/lib/libz.so.1.2.11` or `ldd /usr/bin/bash`. And since `dracut` uses `ldd` to determine the shared library dependencies, any `initramfs` generated by dracut lacks all the required `/usr/lib/*.so*` files. This is the reason why I can't boot with e.g. Void's `initramfs-4.19.32_1.img`.
+ Trying to cross compile for `ppc64` does not work for `build_helper="gir"` packages.I see errors like: `/bin/bash: error while loading shared libraries: /usr/powerpc64-linux-gnu/usr/lib/libdl.so.2: ELF file data encoding not little-endian`. I wonder where I missed to add a case for `ppc64` or if this is an issue with `prelink-cross`.
+ I can however now build packages on the G5 using the cross built `base-chroot`, be it very slow compared to cross building on a Ryzen thread ripper.

So it looks as if only some minor and hopefully solvable problems persist now until I can boot into Void using its own kernel and initrd.img. Next is trying to solve the `ldd` malfunction.

**Comment by pullmoll** [comment]
FYI on my progress with ppc64 on the PowerMac G5.
**tl;dr** things are getting better.

+ The problem with `ldd` was a missing symlink from `/usr/lib` to `/usr/lib64`, because the script has `RTLDLIST="/usr/lib/ld.so.2 /usr/lib64/ld64.so.2"`. Since a ppc64 kernel can also run 32 bit binaries (ppc) I guess this symlink should be created by `glibc` for the `ppc64` case and `ppc` should use `/usr/lib32` - similar to the x86_64 vs. i686 setup.
+ I took the Debian unstable `/boot/config-4.19.0-4-powerpc64` file with just minor changes as `ppc64-dotconfig` for our `linux4.19`, built `4.19.33_1` and can finally boot the G5 with our own kernel and initramfs.
+ Starting X also works, using `nouveau` for the `NV43` I have here, yet I can't build some packages required for `startx` due to the `gir` issue already mentioned. Just too many things depend on `gir` nowadays and I don't want to disable it for `ppc64`. Next I have to identify where this `error … not little-endian` originates from and fix it.

Will now try run linux5.0 on that box and see how it behaves...

**Comment by pullmoll** [comment]
![ppc64-g5-screenshot](https://user-images.githubusercontent.com/10061347/55673703-a3119080-58ab-11e9-9187-63b2aab22058.png)
The first DE I got working is, of course, lxde. It required some more tweaks like making `llvm7` emit `elfv2` also for `ppc64-gnu` and building some packages native. The `build_helper="gir"` packages still don't cross build for me.
